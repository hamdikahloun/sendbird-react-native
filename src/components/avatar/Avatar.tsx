import React, {useState, memo} from 'react';
import {
  GestureResponderEvent,
  Image,
  ImageErrorEventData,
  ImageStyle,
  NativeSyntheticEvent,
  StyleProp,
  StyleSheet,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';

export type AvatarProps = {
  size?: number;
  containerStyle?: StyleProp<ViewStyle>;
  image?: string;
  imageStyle?: StyleProp<ImageStyle>;
  online?: boolean;
  presenceIndicatorContainerStyle?: StyleProp<ViewStyle>;
  onPress?: (event: GestureResponderEvent) => void;
};

const defaultImage =
  'https://avatars.githubusercontent.com/u/5143408?s=200&v=4';

const AvatarMemorized: React.FC<AvatarProps> = props => {
  const {
    containerStyle,
    image,
    imageStyle,
    online,
    presenceIndicatorContainerStyle,
    size = 48,
    onPress,
  } = props;

  const [imageError, setImageError] = useState(false);
  const imageSource = imageError || !image ? defaultImage : image;

  const onError = (error: NativeSyntheticEvent<ImageErrorEventData>) => {
    setImageError(true);
    console.log('Avatar', 'onError', error.nativeEvent.error);
  };

  return (
    <TouchableOpacity disabled={!onPress} onPress={onPress}>
      <View
        style={[
          styles.container,
          {
            borderRadius: size / 2,
            height: size,
            width: size,
          },
          containerStyle,
        ]}>
        <Image
          onError={onError}
          source={{uri: imageSource}}
          style={[
            {
              borderRadius: size / 2,
              height: size,
              width: size,
            },
            imageStyle,
          ]}
          testID={'user-avatar'}
        />
      </View>
      {online && (
        <View
          testID="online-view"
          style={[
            styles.presenceIndicatorContainer,
            presenceIndicatorContainerStyle,
          ]}
        />
      )}
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
    overflow: 'hidden',
  },
  presenceIndicatorContainer: {
    height: 12,
    position: 'absolute',
    right: 0,
    top: 0,
    width: 12,
    backgroundColor: 'green',
    borderRadius: 6,
  },
});

/**
 * Avatar - A round avatar image
 *
 * @example
 *
 * <Avatar image="https://example.com/image.png" size={64} />
 */
export const Avatar = memo(AvatarMemorized);
