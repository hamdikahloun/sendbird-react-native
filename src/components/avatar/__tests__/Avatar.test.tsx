import React from 'react';
import {render, waitFor} from '@testing-library/react-native';
import {Avatar} from '../Avatar';

describe('Avatar', () => {
  it('should render correctly', async () => {
    const screen = render(
      <Avatar image="https://cdn-icons-png.flaticon.com/512/194/194938.png" />,
    ).toJSON();

    expect(screen).toMatchSnapshot();
  });

  it('should render an image with default size', async () => {
    const {queryByTestId} = render(
      <Avatar image="https://cdn-icons-png.flaticon.com/512/194/194938.png" />,
    );

    await waitFor(() => {
      expect(queryByTestId('user-avatar')).toBeTruthy();
    });
  });

  it('should render an image with custom size', async () => {
    const {queryByTestId} = render(
      <Avatar
        image="https://cdn-icons-png.flaticon.com/512/194/194938.png"
        size={20}
      />,
    );

    await waitFor(() => {
      expect(queryByTestId('user-avatar')).toBeTruthy();
    });
  });

  it('should render the default image with wrong image URL', async () => {
    const {queryByTestId} = render(
      <Avatar image="https://example.com/image.png" />,
    );
    await waitFor(() => {
      expect(queryByTestId('user-avatar')).toBeTruthy();
    });
  });

  it('should render the default image with empty URL', async () => {
    const {queryByTestId} = render(<Avatar />);
    await waitFor(() => {
      expect(queryByTestId('user-avatar')).toBeTruthy();
    });
  });

  it('should render the online view', async () => {
    const {queryByTestId} = render(
      <Avatar
        image="https://cdn-icons-png.flaticon.com/512/194/194938.png"
        online
      />,
    );
    await waitFor(() => {
      expect(queryByTestId('online-view')).toBeTruthy();
    });
  });
});
