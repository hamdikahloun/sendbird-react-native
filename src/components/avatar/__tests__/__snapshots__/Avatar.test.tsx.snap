// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should render correctly 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 24,
          "height": 48,
          "width": 48,
        },
        undefined,
      ]
    }
  >
    <Image
      onError={[Function]}
      source={
        Object {
          "uri": "https://cdn-icons-png.flaticon.com/512/194/194938.png",
        }
      }
      style={
        Array [
          Object {
            "borderRadius": 24,
            "height": 48,
            "width": 48,
          },
          undefined,
        ]
      }
      testID="user-avatar"
    />
  </View>
</View>
`;
